### ISSUE TITLE
TestTlogStorageDeadlockWithDeduped is unstable

### ISSUE THREAD

**Issue by GlenDC** [zero-os/0-Disk#231]
`TestTlogStorageDeadlockWithDeduped` fails now and then due to content not matching what it think it should be. This either indicates a bug in the unit-test code, but more likely it probably indicates an async bug in the `tlogStorage` code (or less likely the `dedupeStorage` code).

Log taken from: https://travis-ci.org/g8os/blockstor/jobs/238501436

```
t=2017-06-01T20:48:14+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:42388" module=global caller=redis.go:47
t=2017-06-01T20:48:14+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:38043" module=global caller=redis.go:47
t=2017-06-01T20:48:14+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:35245" module=global caller=redis.go:47
t=2017-06-01T20:48:14+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:46326" module=global caller=redis.go:47
t=2017-06-01T20:48:14+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:43090" module=global caller=redis.go:47
t=2017-06-01T20:48:14+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:36094" module=global caller=redis.go:47
t=2017-06-01T20:48:15+0000 lvl=info msg="embedded LedisDB Server ready and listening at 127.0.0.1:40576" module=global caller=redis.go:47
t=2017-06-01T20:48:15+0000 lvl=info msg="Started listening on local address 127.0.0.1:39700" module=global caller=server.go:55
t=2017-06-01T20:48:15+0000 lvl=dbug msg="received handshake request from incoming connection" module=global caller=server.go:140
t=2017-06-01T20:48:15+0000 lvl=dbug msg="incoming connection checks out with version: 1.1.0-alpha" module=global caller=server.go:150
t=2017-06-01T20:48:15+0000 lvl=dbug msg="create vdisk with expectedSequence:0" module=global caller=vdisk.go:140
t=2017-06-01T20:48:15+0000 lvl=info msg="vdisk a reset first sequence to 0" module=global caller=vdisk.go:210
t=2017-06-01T20:48:15+0000 lvl=dbug msg="handshake phase successfully completed" module=global caller=server.go:175
t=2017-06-01T20:48:15+0000 lvl=dbug msg="replying handshake with status: OK" module=global caller=server.go:202
t=2017-06-01T20:48:15+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:15+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:16+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:16+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:17+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:18+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:18+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:19+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:19+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:20+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:20+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:21+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:21+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:22+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:22+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:23+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:23+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:24+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:24+0000 lvl=dbug msg="vdisk a's tlog server has received force flush message" module=global caller=tlog.go:261
t=2017-06-01T20:48:24+0000 lvl=dbug msg="background context for tlogStorage a aborting" module=global caller=tlog.go:278
t=2017-06-01T20:48:24+0000 lvl=dbug msg="transaction sender for tlogStorage a aborting" module=global caller=tlog.go:309
--- FAIL: TestTlogStorageDeadlockWithDeduped (9.59s)
	ardb_test.go:467: 2 312 unexpected content received [128 138 204 241 121 185 221 143 96 228 222 180 57 217 230 145 175 98 222 35 212 110 152 104 205 169 138 76 254 162 225 210 249 132 84 154 168 183 24 156 218 140 102 84 59 226 159 74 249 118 240 138 45 4 48 42 50 56 98 172 91 23 67 44 241 45 60 132 0 223 169 129 237 203 0 231 162 114 17 56 51 21 190 122 44 57 172 209 28 186 181 152 74 187 59 254 26 248 228 38 55 73 64 32 217 51 19 244 58 28 14 72 47 109 99 179 180 248 7 83 127 96 252 171 66 65 213 137] [96 100 10 109 191 122 90 111 11 188 50 222 33 155 115 249 156 172 91 22 215 104 157 164 93 128 244 135 113 233 122 201 59 53 129 79 151 218 14 190 55 96 38 163 157 81 185 150 211 32 149 125 18 193 115 105 0 150 175 130 33 232 116 214 145 100 180 194 111 113 224 77 94 47 96 144 96 253 169 36 18 178 115 213 71 13 105 19 21 201 22 158 161 129 158 205 71 129 124 137 30 227 238 108 218 203 19 90 36 180 33 4 29 0 165 46 36 60 201 128 97 246 179 213 1 244 183 180]
	ardb_test.go:467: 2 311 unexpected content received [189 88 56 132 105 1 192 186 168 97 171 135 14 226 253 38 183 54 89 133 105 27 224 190 201 82 67 140 92 7 9 7 142 11 248 192 125 142 119 44 97 253 243 87 8 236 208 97 215 30 46 48 58 138 68 0 65 5 136 52 224 77 172 240 79 67 29 187 251 1 89 192 6 75 28 98 230 2 242 170 183 113 220 242 13 203 142 12 55 120 106 179 47 240 64 193 225 25 116 66 237 219 151 35 218 36 201 70 248 32 202 16 71 81 239 151 31 201 192 113 159 108 117 102 24 102 231 84] [114 202 98 85 2 210 171 152 106 104 13 21 46 35 183 145 137 181 156 56 46 65 236 159 50 53 154 71 229 98 153 153 63 68 240 101 211 5 17 7 49 181 18 248 49 6 76 254 251 243 20 180 234 47 104 87 228 246 240 108 233 93 56 105 4 217 193 92 61 31 96 85 201 121 251 57 38 48 73 58 167 23 253 18 170 133 35 38 49 220 111 179 134 227 147 57 31 181 223 70 159 175 134 253 10 207 149 176 44 128 38 53 231 212 252 40 209 244 197 189 211 241 182 221 155 41 76 172]
	ardb_test.go:467: 2 310 unexpected content received [59 183 148 116 147 197 106 167 160 108 161 113 223 147 250 142 237 98 119 145 107 221 157 223 176 22 134 65 25 82 130 63 153 22 24 216 146 22 4 65 213 176 71 203 67 44 33 5 207 189 22 254 207 17 33 11 59 124 21 202 71 4 106 22 171 149 13 127 50 78 108 144 72 91 133 146 208 76 223 58 203 177 181 6 88 252 56 64 127 109 215 228 181 245 62 70 209 180 36 104 214 144 101 248 17 151 66 69 0 244 75 214 91 100 151 134 52 24 141 79 44 129 131 165 64 184 180 247] [224 148 109 181 229 18 182 4 134 232 143 114 216 152 166 244 11 151 122 181 32 134 104 130 159 217 82 149 78 87 84 147 241 237 134 233 201 3 26 16 239 107 240 247 50 152 62 86 242 73 55 20 203 2 180 133 190 220 36 202 168 41 164 151 102 140 58 201 145 194 0 70 86 22 31 222 34 83 177 191 238 149 166 138 179 74 83 120 158 154 67 31 226 108 62 194 168 208 34 187 40 80 138 189 193 223 246 131 120 234 221 159 163 128 147 179 52 99 120 202 254 227 153 89 132 231 127 11]
	ardb_test.go:467: 2 309 unexpected content received [175 5 91 135 76 66 102 56 81 70 127 128 182 34 191 237 102 201 156 239 42 253 221 57 216 114 11 18 116 145 123 120 86 208 95 78 3 8 177 71 65 146 139 11 186 76 98 13 23 18 6 12 201 119 142 205 154 50 224 126 238 73 55 153 162 139 183 161 67 35 48 74 201 133 204 214 66 39 239 146 88 227 58 156 108 135 148 222 191 205 176 249 122 181 128 207 118 196 135 63 237 221 105 225 52 163 107 217 132 85 177 132 123 206 235 134 82 234 251 18 14 16 63 181 91 90 149 120] [164 30 22 145 26 121 158 226 106 127 157 67 192 201 10 2 132 18 200 28 251 85 214 156 196 94 113 172 111 7 150 199 126 39 50 61 144 90 92 234 197 155 69 223 133 164 123 210 34 13 253 86 150 200 174 216 252 51 220 183 116 209 156 133 59 128 184 5 66 45 248 164 127 114 87 250 125 133 9 22 24 179 8 55 89 132 176 126 61 172 57 185 34 47 69 84 200 177 14 94 85 230 183 2 253 196 107 0 190 29 60 2 183 185 108 17 2 136 205 255 160 125 147 131 163 86 92 219]
	ardb_test.go:467: 2 308 unexpected content received [232 154 231 50 82 13 184 14 158 226 226 79 153 125 25 248 253 126 144 49 19 18 106 87 157 90 90 175 42 85 248 82 167 90 182 66 249 222 106 223 42 250 164 142 92 220 218 115 21 8 30 134 207 167 251 197 124 141 31 190 201 59 114 200 234 225 65 238 226 119 222 245 213 51 212 241 217 4 24 236 161 230 30 14 211 21 124 48 8 174 92 227 47 86 107 15 9 122 224 59 241 198 9 214 111 179 200 39 175 230 4 5 83 6 218 80 228 151 129 38 68 34 2 189 233 220 249 144] [18 113 5 136 180 247 117 148 237 2 85 162 143 178 80 171 212 88 253 227 125 77 98 143 117 24 126 164 195 42 5 215 1 150 152 170 9 232 128 80 33 11 161 190 80 84 54 128 192 176 11 228 49 149 116 144 225 148 66 125 15 44 45 60 159 54 47 237 244 170 29 154 148 194 74 79 180 254 24 10 130 37 162 176 44 96 98 185 111 147 10 217 126 248 61 172 150 67 240 71 112 22 130 42 221 136 155 232 77 92 219 167 109 79 154 218 31 102 201 192 115 221 22 54 52 94 30 88]
	ardb_test.go:467: 2 307 unexpected content received [132 241 238 156 153 25 201 19 171 229 84 250 96 42 141 183 179 45 71 133 0 126 172 40 192 235 76 111 232 163 138 183 158 4 233 85 119 242 242 214 190 185 171 71 3 154 148 33 124 222 11 83 230 91 184 97 189 3 84 25 150 10 126 144 250 53 20 129 128 11 38 25 131 250 245 168 48 106 49 174 34 90 117 96 111 234 2 162 219 222 28 150 139 89 129 29 202 147 80 143 128 124 35 199 235 117 98 153 21 234 28 208 78 17 86 53 238 254 189 89 61 49 220 53 194 115 28 66] [99 194 39 215 20 184 103 232 217 67 226 240 50 179 78 78 160 172 134 196 54 12 213 57 201 162 10 35 145 222 0 250 174 238 99 222 29 24 144 50 177 29 146 15 201 191 2 21 231 247 175 149 222 34 112 165 93 114 234 28 95 237 34 47 210 47 105 208 224 173 138 96 116 202 122 151 252 198 59 216 142 211 65 138 121 219 100 242 109 71 50 121 51 83 57 92 90 202 232 110 179 192 2 214 223 203 220 179 255 79 117 43 138 60 68 150 9 57 189 41 214 213 221 64 223 119 238 215]
	ardb_test.go:467: 2 306 unexpected content received [77 229 11 215 110 57 194 240 99 68 63 140 163 43 30 108 117 2 116 111 14 170 165 132 123 20 176 215 213 4 188 175 162 97 127 26 84 6 73 49 190 123 29 241 224 156 10 235 132 3 148 140 94 222 133 132 131 41 246 128 16 14 196 151 38 225 145 185 218 42 103 156 49 18 93 99 133 226 217 243 151 75 119 239 206 26 81 217 107 79 19 6 166 69 17 89 32 241 138 129 163 39 182 120 36 144 105 121 141 62 195 74 42 133 84 78 27 13 191 209 45 183 14 197 147 142 239 124] [68 62 114 205 222 82 13 132 241 151 190 180 174 159 103 18 165 248 198 160 0 148 172 183 49 3 81 92 76 9 167 168 139 46 245 207 125 216 95 190 13 234 104 102 140 54 179 241 56 47 244 17 139 29 33 99 27 60 97 236 210 158 116 79 21 186 3 89 215 39 2 4 167 170 182 115 0 246 227 152 145 141 201 248 226 69 153 188 48 76 123 51 172 246 83 131 35 4 120 59 171 31 79 176 238 250 27 13 164 141 25 179 194 128 102 0 45 238 63 127 229 166 39 49 210 222 204 31]
	ardb_test.go:467: 2 313 unexpected content received [29 76 193 40 161 21 171 155 148 42 196 73 33 151 218 245 61 79 69 111 87 171 18 224 119 236 177 213 27 72 86 231 140 132 59 205 230 220 146 201 78 172 64 70 22 104 219 75 80 234 179 86 98 59 94 94 31 136 53 73 192 4 43 43 175 157 21 139 5 2 42 144 76 227 137 147 254 180 110 202 7 25 193 41 169 193 5 123 123 75 44 108 140 179 236 178 110 174 57 150 234 222 79 107 85 208 201 44 245 107 198 115 135 121 220 230 203 162 206 50 188 17 251 147 83 243 241 180] [70 221 125 173 104 33 215 80 206 68 124 41 220 81 235 144 248 85 65 47 133 7 24 18 81 205 221 229 227 193 12 22 9 179 79 21 60 192 115 177 109 161 78 47 76 230 222 140 137 125 4 42 6 114 156 13 125 117 232 0 149 98 226 180 114 208 158 90 152 104 219 160 246 9 244 213 50 26 12 70 199 193 67 132 70 138 122 49 229 54 40 158 147 174 133 47 173 170 90 122 254 152 50 136 198 23 36 29 42 10 57 49 144 143 52 239 39 206 243 107 52 191 49 147 144 214 73 84]
	ardb_test.go:467: 2 305 unexpected content received [167 98 105 165 167 140 9 94 196 20 82 169 1 201 58 3 101 156 179 115 208 245 129 71 112 209 90 80 147 229 18 90 11 24 194 207 99 42 251 58 28 187 76 190 42 121 225 134 197 134 75 183 203 28 212 86 112 178 73 106 175 246 160 48 133 8 64 211 169 46 81 152 104 97 224 152 31 119 141 176 29 27 122 120 139 205 135 33 11 70 120 34 231 148 196 1 33 128 16 190 124 247 178 169 115 179 30 53 227 226 64 161 166 160 196 28 30 209 190 78 114 109 81 78 151 181 85 205] [172 232 201 118 186 130 242 5 236 169 136 232 48 185 34 86 19 229 179 55 111 251 128 117 247 125 233 96 164 89 73 121 103 120 19 77 225 142 123 93 91 155 229 119 36 152 13 46 67 44 183 176 233 29 214 116 31 206 222 95 39 93 32 217 226 50 167 98 67 179 51 177 67 231 130 81 0 60 83 85 102 68 63 77 183 219 213 125 17 250 154 63 170 68 207 62 143 100 95 156 139 120 116 179 210 94 171 0 231 231 182 127 70 222 70 169 1 35 44 78 56 221 18 163 209 107 74 76]
	ardb_test.go:467: 2 314 unexpected content received [47 157 93 61 93 124 191 32 100 209 121 44 30 46 172 3 238 69 115 197 92 136 132 25 43 225 223 193 66 76 15 229 203 105 238 221 224 228 246 104 224 182 186 230 178 63 27 246 215 163 146 155 61 5 247 175 140 224 37 241 199 193 34 69 88 73 186 36 252 3 202 179 196 78 40 204 8 56 150 102 44 174 49 124 171 167 249 240 176 193 139 68 204 52 148 52 131 96 40 175 136 98 236 94 188 205 61 65 156 2 42 240 140 226 35 224 84 45 3 76 121 22 169 117 102 22 251 62] [236 3 46 221 255 150 179 50 73 200 156 253 50 139 225 184 64 39 72 144 69 18 150 53 8 107 187 195 123 116 179 39 136 216 150 191 85 68 92 211 77 196 162 40 51 89 218 0 186 36 36 76 18 81 173 125 207 217 69 72 167 63 30 96 212 169 53 251 95 224 126 76 91 202 84 125 223 205 118 152 94 186 178 195 207 225 19 69 92 56 20 226 245 197 65 186 247 70 176 62 99 224 143 182 38 19 50 61 137 56 199 97 2 204 32 250 83 222 103 64 33 32 206 231 9 166 75 204]
t=2017-06-01T20:48:24+0000 lvl=info msg="connection from 127.0.0.1:55642 dropped" module=global caller=server.go:105
t=2017-06-01T20:48:24+0000 lvl=info msg="Started listening on local address 127.0.0.1:33305" module=global caller=server.go:55
t=2017-06-01T20:48:24+0000 lvl=dbug msg="abort current sendResp goroutine for vdisk:a" module=global caller=server.go:287
```

**Comment by GlenDC** [comment]
Failed again on master in https://travis-ci.org/zero-os/0-Disk/jobs/239111363, so still an active issue.

**Comment by GlenDC** [comment]
@robvanmieghem @grimpy not sure if you were already testing tlog stuff with the milestone alpha-3 build of `zero-os/0-Disk`. if so, you might want to postpone that until the milestone alpha-4 build, or by simply using master, or by using the version as it as at commit 42e472727ca3acb4cfaa1434b688ab84272de687.

Because 42e472727ca3acb4cfaa1434b688ab84272de687 turns out to be resolving a pretty critical async flaw, that was present in the alpha-3 build, which can give the issues that the logs mentioned in this issue expose.

More information about the issue and the solution to this issue can be found in the description of 42e472727ca3acb4cfaa1434b688ab84272de687.

**Comment by GlenDC** [comment]
Merged PR #240 also fixes some data races, detected as warnings using `go test -race`. I don't think these will really cause serious issues in case you were already testing, as they never seemed to have given any problems in our manual tests. But you know, the warnings were there, and have since today been resolved. Data race tests are now also part of the travis test suite, so should help detect them earlier from now on.
